Implement SnowflakeConnection

vdonato opened the pull request:
This PR implements `SnowflakeConnection`, which is intended to replace the existing `SnowparkConnection`
class that we ship as a default connection. This new class is a more general connection, but it also retains all
of the functionality that the existing `SnowparkConnection` class does as it allows users to create Snowpark
sessions from the underlying connection object using the `.session()` helper method.

vdonato committed:
Tweak missing dependency warnings for new SnowflakeConnector

vdonato committed:
Add skeleton SnowflakeConnection

vdonato committed:
Add skeleton of SnowflakeConnection as a first party connection

vdonato committed:
Take first pass at implementing SnowflakeConnection

vdonato committed:
Fix some type errors and typos

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,217 @@
+        schema: str | None = None,
+        chunk_size: int | None = None,
+        **kwargs,
+    ) -> tuple[  # This return type is somewhat horrifying but is the return type of pandas_tools.write_pandas, so we keep it consistent here.
+        bool,
+        int,
+        int,
+        Sequence[
+            tuple[
+                str,
+                str,
+                int,
+                int,
+                int,
+                int,
+                str | None,
+                int | None,
+                int | None,
+                str | None,
+            ]
+        ],
+    ]:
```
Note to self: Take a look at whether we want to just return a subset of these values for the Streamlit version

sfc-gh-jcarroll replied:
@vdonato here's the current return value for a simple insert of 3 rows:

```python
import streamlit as st
import pandas as pd

conn = st.connection("snowflake")
conn.cursor().execute("CREATE TABLE IF NOT EXISTS MY_TABLE_FOO(IDENT NUMBER, AMOUNT NUMBER)")
df = pd.DataFrame([[1, 2],[3, 4], [5, 6]], columns=["IDENT", "AMOUNT"])
res = conn.write_pandas(df, "MY_TABLE_FOO")
st.write(res)
```
`(True, 1, 3, [('xxxxx/file0.txt', 'LOADED', 3, 3, 1, 0, None, None, None, None)])`

In Snowflake example docs for the function, they even drop that last sequence/tuple:
`success, nchunks, nrows, _ = write_pandas(cnx, df, 'customers')`

Why don't we follow the same pattern and just return a 3 element tuple `success, nchunks, nrows,` (I would almost argue that we reduce it down further to just return `nrows` and do `None` if `success==False`, but maybe that's too simplified)

vdonato replied:
Yeah, I'm in favor of just retaining the first three elements of the tuple for this one

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,217 @@
+        """
+        from snowflake.snowpark.session import Session
+
+        return cast(
+            Session, Session.builder.configs({"connection": self._instance}).create()
+        )
```
And we believe that this version will be thread safe, right?

vdonato replied:
Fairly sure -- the `Session` object itself not being thread safe should be circumvented by creating a new session object each time this method is called

There is some potential for a user to save a Snowpark session in session state or something like that and run into issues, but the pattern would be quite unnatural if you're using `st.connection` already

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,217 @@
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.connections.util import load_from_snowsql_config_file, running_in_sis
```
Note to snowflake drivers reviewers: Worth taking a look at the implementations of these methods too (outside this PR) and seeing if you agree - in particular `running_in_sis()` is important!

sfc-gh-mkeller replied:
For `running_in_sis` you should use the following snowpark utility method instead https://github.com/snowflakedb/snowpark-python/blob/697294d8ee89396deab4309dea384b1eadb75d6b/src/snowflake/snowpark/_internal/utils.py#L500-L501
We're in the process of renaming the SP connection to match the public one.

vdonato replied:
👍 will do

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,217 @@
+            )
+
+        # If we have a version of the Snowflake Python Connector installed that predates
+        # the ConfigManager, look for credentials in  ~/.snowsql/config as a last
```
With snowcli adoption increasing, check if this is still the (only) path we want to check or not. Don't want to over-complicate this logic though!!

Wonder if we should call out in the docs somewhere "this connection works best with snowflake driver version >= x.x.x" based on when ConfigManager was added

sfc-gh-bwarsaw replied:
> this connection works best with snowflake driver version >= x.x.x

Can you lower pin a dependency to enforce this?

sfc-gh-jcarroll replied:
@vdonato after thinking about this a bit more, I'm in favor of just removing the approximated snowsql support from the SnowflakeConnection. I think it made sense in the spring, but now that ConfigManager support has been out for 3 months and snowcli workflows are improving, it isn't really needed and it just adds maintenance burden / crufty code.

Problems with what we have now:
- It doesn't actually work for Windows
- It only looks in one of the expected files on Linux, not all of them
- We don't really even document that it works, just a casual mention in the tutorial which suggests that we have the full support (not really true)

We can leave the current implementation in the SnowparkConnection which will be deprecated, but I think we just rip this part out in this updated version.

Then the story is clearer and easier to document explicitly:
- using snowflake-connector-python >= 3.1.0 - you can use Streamlit secrets or the [connections.toml file](https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-example#connecting-using-the-connections-toml-file) which is fully supported and documented, + explicit kwargs
- using an earlier version - you can only use Streamlit secrets or explicit kwargs

WDYT?

vdonato replied:
yep -- this sounds reasonable to me. I'll get rid of the code that attempts to load credentials from the snowsql config file

vdonato commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,217 @@
+        # have available.
+
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+            # the kwargs passed to `st.connection`.
+            conn_kwargs = {**st_secrets, **kwargs}
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            # TODO(vdonato): Double-check that we want to map the "snowflake" connection
+            # name in a call to `st.connection` to the "default" connection in the
+            # Snowflake config file. This is aligned with the default behavior for the
+            # underlying python connector and seems more sane than having a "snowflake"
+            # connection in a snowflake-specific config file.
```
btw @sfc-gh-jcarroll these two `TODO`s could use a bit of product input.

I'm fairly certain that both implementations here are pretty reasonable (I think we'll want `kwargs` to apply to any method of creating a `SnowflakeConnection`, and mapping `snowflake` -> `default` seems the most natural to me), but it's worth getting some input on this

vdonato replied:
The other thing to consider is whether we want to map the `snowflake` connection name to `default` as well on L93

sfc-gh-jcarroll replied:
Yeah for the above 2 TODOs that looks correct to me as well. We should get confirmation from Mark that it seems consistent to him too.

For L93 I'm not totally sure I follow, maybe we can discuss on Monday.

vdonato committed:
Add unit tests and finish writing docstrings

vdonato commented on lib/tests/streamlit/connections/snowpark_connection_test.py:
```diff
@@ -133,8 +133,6 @@ def test_retry_behavior(self):
         MagicMock(),
     )
     def test_retry_fails_fast_for_most_errors(self):
-        from snowflake.snowpark.exceptions import SnowparkServerException
-
```
unrelated change to remove an unused import, but I don't think it's worth creating a separate PR for this

sfc-gh-mkeller commented on lib/streamlit/connections/__init__.py:
```diff
@@ -14,6 +14,9 @@
 # Explicitly re-export public symbols.
 from streamlit.connections.base_connection import BaseConnection as BaseConnection
+from streamlit.connections.snowflake_connection import (
+    SnowflakeConnection as SnowflakeConnection,
```
Just curious why you name it the same thing this way?

sfc-gh-bwarsaw replied:
...and below.  I just looked at the original source and this seems to be a common pattern.  Is this done for linting purposes?  Because it really isn't necessary for re-export, and definitely seems odd.  If you're doing it to avoid some linter warnings, it might be better to just explicitly ignore the linter warnings through file pragmas, which at least would be self explanatory.  E.g. We ignore [F401](https://github.com/snowflakedb/snowapi-python/blob/main/ruff.toml#L27) in snowapi-python.

vdonato replied:
I believe this is needed by `mypy` so that it knows which parts of a module should be considered part of the public interface.

I'll punt on making changes to this in this PR since there are other places in the codebase that would need to be changed if we are able to get rid of it (although I think we'll need to keep it around for the typechecker), but I'll either see whether we can get rid of it or will improve the comments around these reexports in a followup change

sfc-gh-bwarsaw replied:
A `__all__` should do the trick for mypy IIRC.  What doesn't work yet in mypy is [computed dunder-all](https://github.com/python/mypy/issues/15300).  It makes sense to punt to a future PR.

vdonato replied:
ah, I was unaware -- thanks! I'll address this throughout the codebase in a future PR to `develop`

sfc-gh-mkeller commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, Dict, Sequence, cast
```
You're using future annotations, instead of `typing.Dict` you should just use `dict`
```suggestion
from typing import TYPE_CHECKING, Any, Sequence, cast
```

sfc-gh-bwarsaw replied:
Pedant Trigger Alert 😆 

Future annotations don't directly permit the use of `dict` instead of `typing.Dict` but it has that effect because PEP 563 [defers annotation evaluation](https://peps.python.org/pep-0563/) at function definition time, and instead stores them as strings.  `typing.Dict` is deprecated [as of Python 3.9](https://docs.python.org/3/library/typing.html#typing.Dict) in favor of `dict`, but I think Python 3.8 is probably still your minimum version.

So Mark is correct; since you're using future annotations, you might as well save yourself an import.  And "lucky" for us, [PEP 649](https://peps.python.org/pep-0649/) will take over the world as of Python 3.13 and beyond.

sfc-gh-mkeller commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+            # underlying python connector and seems more sane than having a "snowflake"
+            # connection in a snowflake-specific config file.
+            connection_name = (
+                "default"
```
You can also retrieve the default connection name from the CONFIG_MANAGER like this:
```suggestion
                snowflake.connector.connection.CONFIG_MANAGER["default_connection_name"]
```

sfc-gh-mkeller replied:
Note that this was added in version 3.2.0

sfc-gh-jcarroll replied:
I was playing with this, I think we should check for CONFIG_MANAGER["default_connection_name"] if available and then fall back to "default" as the value otherwise.

It seems to work nicely, I was able to do `snow connection add` with snowcli, create a connection named "default", and then `st.connection("snowflake")` just worked. Same for connections of other types. This is really slick!!!

sfc-gh-jcarroll replied:
also since the default connection name value comes from `SNOWFLAKE_DEFAULT_CONNECTION_NAME`, I think it means a developer could just add that to their secrets.toml to point to whatever connection they have defined, and just use `st.connection("snowflake")`. This might be a more common pattern for SiS apps being developed locally.

sfc-gh-bwarsaw commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.connector import (  # type: ignore # isort: skip
```
Rather than generic `type: ignore` pragmas, you might consider [including the error code](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html?highlight=ignore#type-ignore-error-codes) explicitly.  It helps with readability.

vdonato replied:
ack -- will do

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+            "cache miss" and the cached resource is being created. If a string, the value
+            of the show_spinner param will be used for the spinner text.
+        params : list, tuple, dict or None
+            List of parameters to pass to the execute method. The syntax used to pass
+            parameters is database driver dependent. Check your database driver
+            documentation for which of the five syntax styles, described in `PEP 249
+            paramstyle <https://peps.python.org/pep-0249/#paramstyle>`_, is supported.
+            Default is None.
```
Need to update this with the snowflake specific docs and behavior. This might be the best cite: https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-example#label-python-connector-binding-data

sfc-gh-jcarroll replied:
Probably worth noting the default / recommended style and including an example somewhere too

Seems like both the client side approaches work on local

```python
conn = st.connection("snowflake")
# Broken, see comment below
# df = conn.query("SELECT %(col1)s, %(col2)s", ttl=0, params={'col1': 789, 'col2': 'test string3'})
df = conn.cursor().execute("SELECT %s, %s", (789, 'test string3')).fetch_pandas_all()
df2 = conn.cursor().execute("SELECT %(col1)s, %(col2)s", {'col1': 123, 'col2': 'foo bar'}).fetch_pandas_all()
st.dataframe(df)
st.dataframe(df2)
```

qmark doesn't seem to work by default (which I think matches the documentation). Guessing that's OK for now.

vdonato replied:
ack -- will punt this to a followup PR along with other potential tweaks to docstrings

sfc-gh-bwarsaw commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+
+
+def _validate_connection_params(params: Dict[str, Any]) -> None:
+    for p in _REQUIRED_CONNECTION_PARAMS:
```
You might consider something like (untested):

```python
   missing = _REQUIRED_CONNECTION_PARAMS - set(params)
   if len(missing) > 0:
      raise StreamlitAPIException(f"Missing Snowflake connection params: {', '.join(missing)}")
```

which prints all the missing required connection parameters, if say in the future you add more required ones to the above set.

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+        )
+        def _query(sql: str) -> pd.DataFrame:
+            cur = self._instance.cursor()
+            cur.execute(sql, **(params or {}))
```
I think this is incorrect - params value should be passed as `parameters` arg, not expanded as kwargs. I get an error when I try to use it:

```python
df = conn.query("SELECT %(col1)s, %(col2)s",
                ttl=0,
                params={'col1': 789, 'col2': 'test string3'})
```

`TypeError: SnowflakeCursor.execute() got an unexpected keyword argument 'col1'`

https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#id2

sfc-gh-jcarroll replied:
(and also get an error when passing a tuple and using qmark syntax)

vdonato replied:
Whoops, I misunderstood what the notation in the docs meant -- will fix

sfc-gh-bwarsaw commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+            raise StreamlitAPIException(f"Missing Snowflake connection param: {p}")
+
+
+class SnowflakeConnection(BaseConnection["InternalSnowflakeConnection"]):
```
It sure feels like we want to refactor all these connection patterns into a more base library, even if we can't do that now.  @sfc-gh-mkeller what do you think about coming up with a plan to eventually do that?  Seems like a much better long term solution than having all these separate connection classes.

sfc-gh-bwarsaw commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+
```
Nit: extra blank line?

vdonato replied:
ah, yeah -- although I think the TODO is no longer needed so can just be removed entirely

sfc-gh-bwarsaw commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+        schema: str | None = None,
+        chunk_size: int | None = None,
+        **kwargs,
+    ) -> tuple[  # This return type is somewhat horrifying but is the return type of pandas_tools.write_pandas, so we keep it consistent here.
```
Pandas doesn't have a type alias for this mess? 😞

vdonato replied:
Doesn't seem like it given this type signature is just copied from the function we're wrapping 😞

Thankfully, it looks like we're going to be simplifying this type in the helper function to just return the first 3 elements of the tuple given how rarely the last element is used, so that simplifies this by a good amount

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+        from snowflake.snowpark.session import Session
+
+        return cast(
+            Session, Session.builder.configs({"connection": self._instance}).create()
```
This breaks when you run in SiS today - will post the full code in slack

`SnowparkSessionException: (1410): In a stored procedure, you shouldn't create a session. The stored procedure provides a session.`

Guess we need to either track the session we get during init, or check if we're in SiS and do get_active_session() again here? Not sure how we get around the thread safety issue in that case 🤔 cc @sfc-gh-mkeller

vdonato replied:
I think thread-safety is thankfully a non-issue in SiS because every browser tab gets its own Streamlit runtime, so I think we should be fine just calling `get_active_session` here

sfc-gh-jcarroll replied:
👍

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+            Enable the spinner. The default is to show a spinner when there is a
+            "cache miss" and the cached resource is being created. If a string, the value
+            of the show_spinner param will be used for the spinner text.
+        params : list, tuple, dict or None
```
As noted in Slack, when running outside of SiS I think we need to enable the qmark binding style by default, so code works across both. This is also consistent with the snowpark params support ([cite](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/api/snowflake.snowpark.Session.sql))

vdonato replied:
👍

sfc-gh-jcarroll commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,260 @@
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+
+            # the kwargs passed to `st.connection`.
+            conn_kwargs = {**st_secrets, **kwargs}
+            _validate_connection_params(conn_kwargs)
+            return snowflake.connector.connect(**conn_kwargs)
```
It would be nice if I can do this:

```toml
# .streamlit/secrets.toml

[connections.snowflake]
connection_name = "my-connection-name-from-snowcli"
```

I *think* it would just need to pass the validation for that to work. Right now it fails with `StreamlitAPIException: Missing Snowflake connection param: account`. Might be worth adding an exception, WDYT?

vdonato replied:
Hm, I'll punt on the decision for this one until Monday (I'm out early in the afternoon tomorrow) / potentially in a small followup PR. This seems reasonable to me, but if we have to implement logic to accept either a `connection_name` or an `account`, then I think it may be more reasonable to just not validate any parameters and allow the underlying library to throw the error in case a user did something wrong.

This won't be quite as nice as the more friendly error message that we provide right now, but the amount of extra handling to implement something that already exists one layer down may not be worth it for anything more complex than very simple error checks

vdonato committed:
Address non-behavior-related PR comments

vdonato committed:
Last few SnowflakeConnection tweaks

vdonato commented:
Thanks for the reviews @sfc-gh-bwarsaw @sfc-gh-mkeller!

I think I've addressed most of the PR comments aside from a few that we can punt to a future PR, so it may be useful to give the relevant parts of this PR another quick read

vdonato committed:
Add missing expected param to test

lukasmasuch commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,252 @@
+            session = get_active_session()
+
+            if hasattr(session, "connection"):
+                return session.connection
```
nit: Why is there sometimes `session.connection` and `session._conn._conn`? Is it related to different versions? Maybe add a comment here to explain this aspect

vdonato replied:
Will do -- I'll go ahead and click merge on this PR to avoid having to wait on another CI run, but I'll add these comments in a followup PR for docstring tweaks and comment adjustments

lukasmasuch commented on lib/streamlit/connections/snowflake_connection.py:
```diff
@@ -0,0 +1,252 @@
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            config_mgr = snowflake.connector.connection.CONFIG_MANAGER
```
nit: is the `CONFIG_MANAGER` also something that was added in more recent versions of the snowflake connector?

vdonato replied:
Yeah -- this is only available in more recent versions -- I'll add an explanatory comment in a followup PR

lukasmasuch reviewed (approved):
LGTM 👍

vdonato commented:
Going to go ahead and click merge on this one since it's just to get the code into the feature branch, and none of the `st.connection` code will land on `develop` until #7536 is checked in. There are some minor tweaks that we still want to make to address feedback here that we'll make in a followup PR into the feature branch
